### PR TITLE
Rick/DPC-241 refactor aggregation engine

### PR DESCRIPTION
**Pull request by RickHawesUSDS**
**Why**

This pull-requests implements three changes needed for R3.

- The output files are compatible with the lastest FHIR Bulk Spec [https://github.com/HL7/bulk-data/blob/master/spec/export/index.md](FHIR Bulk Data Export Implementation Guide)
- Export jobs are more performant by taking advantage of multiple cores to do parallel fetches.
- Large export jobs are handled better by paginating BlueButton request. This avoids long requests to BB which may timeout. The output files are broken up into multiple files as well to avoid timeouts from the client.

**What Changed**

- The AggregatorEngine and EncryptingAggregatorEngine has been restructured into 1 main class and 3 helper classes. Encryption is now a feature flag on the pipeline, instead of an override of the main class.
- Multiple RxJava pipelines is now merged together into a single pipeline for a job.
- The pipeline uses thread-safe structures. JobModel and JobResult are now immutable structures.
- There is one JobResult per output file instead one per resource type. JobResult has a sequence field now.
- Within a job, the resources are requested from BlueButton in parallel (one fetch per core). Parallel fetching is controlled by a config flag which is turned off by default.
- New aggregator configs: parallelRequestsEnabled, resourcesPerFile and resourcesPerRequest. 

**Choices Made**

- The decision to unbundle resources was made to follow the new export spec and to match BCDA precedent.
- The decisions to refactor the engine was made to pass CodeClimate checks.
- The choice to use Flowables was driven by the .parallel operators that it offered.
- The choice to remove encryption parameters from the config file was made based on the view that encryption parameters were an engineering decision (ie. its govern by many NIST recommendations) and not something that we want to alter in production. 

**Tickets Closed**

- DPC-171 - BlueButton client should handle paginated responses
- DPC-231 - As a developer, I want to be able to disable file encryption for testing purposes
- DPC-241 - Refactor Aggregation Engines
- DPC-145 - AggregationEngine should dispatch BB requests on its own thread pool.
- DPC-234 - As an operator, I would like to DPC to page their output files for very large (> 10K patient) providers

**Future Work**

Performance work should start.
Turn on parallel requests after some more experience with the code.
Encryption metadata should be included per file as an extension (or a BCDA keymap needs to be provided). The JobResult structure can be used to store this information.

### CODE REVIEW

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java [hunk 1]
```diff
@@ -1,66 +1,67 @@
...
      * @param bbclient    - {@link BlueButtonClient } to use
      * @param queue       - {@link JobQueue} that will direct the work done
-     * @param context     - {@link FhirContext} for DSTU3 resources
+     * @param fhirContext - {@link FhirContext} for DSTU3 resources
      * @param exportPath  - The {@link ExportPath} to use for writing the output files
      * @param retryConfig - {@link RetryConfig} injected config for setting up retry handler
      */
     @Inject
-    public AggregationEngine(BlueButtonClient bbclient, JobQueue queue, FhirContext context, @ExportPath String exportPath, RetryConfig retryConfig) {
+    public AggregationEngine(BlueButtonClient bbclient, JobQueue queue, FhirContext fhirContext, @ExportPath String exportPath, Config config, RetryConfig retryConfig) {
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java]: Makes sense, but CodeClimate will complain about the number of parameters. I'll guess I'll create a sub config structure.

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/CipherBuilder.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     * "PublicKey" : (String) Base64-encoded RSA public key that was initially provided by the vendor
+     * }
+     * }
+     *
+     * @param rsaPublicKey - the public key used encrypt symmetric key used for encryption
+     * @return JSON formatted string
+     * @throws GeneralSecurityException for config errors
+     * @throws IOException for JSON errors
+     */
+    String getMetadata(byte[] rsaPublicKey) throws GeneralSecurityException, IOException {
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/CipherBuilder.java]: Yes. That is the way Issac implemented it. It's wasn't intended to be this way in the end.

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/ResourceFetcher.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * Given a bundle, return a list of resources in the passed in bundle and all
+     * the resources from the next bundles.
+     *
+     * @param patientID to fetch for
+     * @param firstBundle of resources. Included in the result list
+     * @return a list of all the resources in the first bundle and all next bundles
+     */
+    private List<Resource> fetchAllBundles(String patientID, Bundle firstBundle) {
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/ResourceFetcher.java]: I think the construct that you want is a generator. Resiliancy4j didn't work with generators. Couldn't find the actual bug, but spent a lot of time looking.

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java [hunk 4]
```diff
@@ -133,171 +113,84 @@ private void workExportJob(Pair<UUID, JobModel> workPair) {
...
-        } else if (ex instanceof BaseServerResponseException) {
-            final var serverException = (BaseServerResponseException) ex;
-            details = String.format("Blue Button error: HTTP status: %s", serverException.getStatusCode());
+        // Make this flow hot (ie. only called once)
+        final var fetcher = new ResourceFetcher(bbclient, retryConfig, job.getJobID(), resourceType);
+        final Flowable<Resource> mixedFlow;
+        if (parallelRequests) {
+            mixedFlow = Flowable.fromIterable(job.getPatients())
+                    .parallel()
+                    .runOn(Schedulers.io())
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java]: Parallel limits to to number of cpus. So it won't starve other activities by creating too many threads. Added DPC-301.

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java [hunk 5]
```diff
@@ -133,171 +113,84 @@ private void workExportJob(Pair<UUID, JobModel> workPair) {
...
-                .setLocation(location);
-        return outcome;
+        final var connectableMixedFlow = mixedFlow.publish().autoConnect(2);
+
+        // Batch the non-error resources into files
+        final var counter = new AtomicInteger();
+        final var writer = new ResourceWriter(fhirContext, config, exportPath, job, resourceType);
+        final Flowable<JobResult> resourceFlow = connectableMixedFlow.filter(r -> r.getResourceType() != ResourceType.OperationOutcome)
+                .buffer(resourcesPerFile)
+                .map(batch -> writer.writeBatch(counter, batch));
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/AggregationEngine.java]: Right. DPC-301 captures this work.

File: dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/ResourceFetcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickrobison-usds) [reply dpc-aggregation/src/main/java/gov/cms/dpc/aggregation/engine/ResourceFetcher.java]: Perhaps it's time to abandon the resilience4j plan. Seems like we can replicate the functionality using only rxJava primitives. Plus, that will avoid requiring us to pull in a huge external dependency (Vavr). Seems like a future task though.

**Pull request merged**